### PR TITLE
DATAREDIS-1151 - change clean method  `keys` command to `scan`

### PR DESCRIPTION
**Pull request by eyison**
`clean` method use `keys` command to get all keys which is blocked , may cause other commands time out.

https://jira.spring.io/browse/DATAREDIS-1151

### CODE REVIEW

**Comment by pivotal-issuemaster** [comment]
@eyison Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-data-redis&pullRequestId=532)!

[Click here](https://cla.pivotal.io/sync/spring?repositoryId=spring-projects/spring-data-redis&pullRequestId=532) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

**Comment by pivotal-issuemaster** [comment]
@eyison Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-data-redis&pullRequestId=532)!

File: src/main/java/org/springframework/data/redis/cache/DefaultRedisCacheWriter.java [hunk 1]
```diff
@@ -179,11 +179,14 @@ public void clean(String name, byte[] pattern) {
...
-				if (keys.length > 0) {
-					connection.del(keys);
+				List<byte[]> keys = new ArrayList<byte[]>();
+				Cursor<byte[]> cursor = connection.scan(new ScanOptions.ScanOptionsBuilder().match(new String(pattern)).build());
+				while (cursor.hasNext()) {
+					byte[] bytes = cursor.next();
+					keys.add(bytes);
+				}
+				if (!keys.isEmpty()) {
+					connection.del(keys.toArray(new byte[][]{}));
```
Comment (chengchen) [inline src/main/java/org/springframework/data/redis/cache/DefaultRedisCacheWriter.java:189]: As the complexity of DEL is O(N), if the number of keys is huge, DEL command could time out as well, right?
Maybe we could do it batch by batch as well?

I was thinking of using [UNLINK](https://redis.io/commands/unlink), but it's only supported in higher Redis version and it's a change of behavior (from sync to async)...

File: src/main/java/org/springframework/data/redis/cache/DefaultRedisCacheWriter.java [hunk 2]
```diff
@@ -179,11 +179,14 @@ public void clean(String name, byte[] pattern) {
 					wasLocked = true;
 				}
 
-				byte[][] keys = Optional.ofNullable(connection.keys(pattern)).orElse(Collections.emptySet())
-						.toArray(new byte[0][]);
-
-				if (keys.length > 0) {
-					connection.del(keys);
+				List<byte[]> keys = new ArrayList<byte[]>();
```
Comment (cnhongwei) [inline src/main/java/org/springframework/data/redis/cache/DefaultRedisCacheWriter.java]: Use LinkedList replace ArrayList because the list may be huge.

**Comment by mp911de** [comment]
The change performs a `DEL` call per cache key which isn't efficient. We have received a newer PR with #547 that supersedes this one.

**Pull request closed without merging**